File: app/src/main/java/com/andysklyarov/finnotify/framework/ApplicationViewModelFactory.java
package com.andysklyarov.finnotify.framework;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.andysklyarov.finnotify.interactors.Interactors;
import com.andysklyarov.finnotify.presentation.MainViewModel;

import java.lang.reflect.InvocationTargetException;

    public final class ApplicationViewModelFactory extends ViewModelProvider.NewInstanceFactory   {

    @NonNull
    private final Application application;
    private static Interactors useCase;

    public ApplicationViewModelFactory(@NonNull Application application) {
        this.application = application;
    }

    static void inject(Interactors newUseCase) {
        useCase = newUseCase;
    }


    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {

        if (modelClass == MainViewModel.class) {
            try {
                return modelClass.getConstructor(Application.class, Interactors.class).newInstance(application, useCase);
            } catch (IllegalAccessException | InstantiationException | InvocationTargetException | NoSuchMethodException e) {
                e.printStackTrace();
            }
        }

        throw new IllegalStateException("ViewModel must extend MainActivityViewModel");
    }
}

File: app/src/main/java/com/andysklyarov/finnotifyfree/framework/ApplicationViewModelFactory.kt
package com.andysklyarov.finnotifyfree.framework

import android.app.Application
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.andysklyarov.finnotifyfree.interactors.Interactors
import com.andysklyarov.finnotifyfree.presentation.MainViewModel
import java.lang.reflect.InvocationTargetException

class ApplicationViewModelFactory(private var application: Application) :
    ViewModelProvider.NewInstanceFactory() {

    companion object {
        private lateinit var useCase: Interactors
        fun inject(newUseCase: Interactors) {
            useCase = newUseCase
        }
    }

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass == MainViewModel::class.java) {
            try {
                return modelClass.getConstructor(Application::class.java, Interactors::class.java)
                    .newInstance(application, useCase) as T
            } catch (e: IllegalAccessException) {
                e.printStackTrace()
            } catch (e: InstantiationException) {
                e.printStackTrace()
            } catch (e: InvocationTargetException) {
                e.printStackTrace()
            } catch (e: NoSuchMethodException) {
                e.printStackTrace()
            }
        }
        throw IllegalStateException("ViewModel must extend MainActivityViewModel")
    }
}